Plugins/AutocompleteManager/AutocompleteManager.swift

```diff
@@ -0,0 +1,342 @@
+    }
+    
+    /// Invalidates the `currentSession` session if it existed
+    private func unregisterCurrentSession() {
+        
+        guard let session = currentSession else { return }
+        guard delegate?.autocompleteManager(self, shouldUnregister: session.prefix) != false else { return }
+        currentSession = nil
+        delegate?.autocompleteManager(self, shouldBecomeVisible: false)
+    }
+    
+    /// Calls the required methods to relayout the `AutocompleteTableView` in it's superview
+    private func layoutIfNeeded() {
+        
+        tableView.reloadData()
+        
+        // Resize the table to be fit properly in an `InputStackView`
+        tableView.invalidateIntrinsicContentSize()
+        
+        // Layout the table's superview
+        tableView.superview?.layoutIfNeeded()
+    }
+    
+}
+
+extension AutocompleteManager: UITextViewDelegate {
+    
+    // MARK: - UITextViewDelegate
+    
+    public func textViewDidChange(_ textView: UITextView) {
+        reloadData()
+    }
+    
+    public func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+        
+        // Ensure that the text to be inserted is not using previous attributes
+        preserveTypingAttributes()
+        
+        // range.length > 0: Backspace/removing text
+        // range.lowerBound < textView.selectedRange.lowerBound: Ignore trying to delete
+        //      the substring if the user is already doing so
+        if range.length > 0, range.lowerBound < textView.selectedRange.lowerBound {
+            
+            // Backspace/removing text
+            let attribute = textView.attributedText
+                .attributes(at: range.lowerBound, longestEffectiveRange: nil, in: range)
+                .filter { return $0.key == NSAttributedAutocompleteKey }
+            
+            if (attribute[NSAttributedAutocompleteKey] as? Bool ?? false) == true {
+                
+                // Remove the autocompleted substring
+                let lowerRange = NSRange(location: 0, length: range.location + 1)
+                textView.attributedText.enumerateAttribute(NSAttributedAutocompleteKey, in: lowerRange, options: .reverse, using: { (_, range, stop) in
+                    
+                    // Only delete the first found range
+                    defer { stop.pointee = true }
+                    
+                    let emptyString = NSAttributedString(string: "", attributes: typingTextAttributes)
+                    textView.attributedText = textView.attributedText.replacingCharacters(in: range, with: emptyString)
+                    textView.selectedRange = NSRange(location: range.location, length: 0)
+                })
+                unregisterCurrentSession()
+                return false
+            }
+        }
+        return true
+    }
+    
+}
+
+extension AutocompleteManager: UITableViewDelegate, UITableViewDataSource {
+    
```

Comment:
It may be a better approach to stop using this pattern of confirming to a protocol via extension. This doesn't allow users to override any of the methods declared here

Reply:
Yup I agree